README.md
[![Clojars Project](http://clojars.org/djy/latest-version.svg)](http://clojars.org/djy)

# What's this?

This is a library of character utility functions for Clojure, inspired by useful
built-in string and character libraries from other languages, most significantly
Haskell's Data.Char library.

It is currently somewhat cumbersome to work with characters in Clojure.
Complicating matters is the inherent complexity of dealing with supplementary
characters in the JVM; Java characters are 16-bit, allowing characters in the
Unicode range 0000-FFFF to be expressed as single characters. This range is
called the Basic Multilingual Plane (BMP), however the range of existent
characters has since expanded, bringing about the need for 32-bit characters.
Java's way of representing these supplementary characters is via pairs of 16-bit
characters, for a combined total of 32 bits.

This library aims to provide convenient wrappers for standard Java Character
library functions, as well as some new utility functions to facilitate working
with characters.

Many of these functions are polymorphic in nature, by way of a `HasCodePoint`
protocol exposing a `code-point-of` function, which can take as an argument a
character, an integer representing a Unicode code point, or a string beginning
with a supplementary character (i.e. two 16-bit Java characters). This allows
us to work with BMP and supplementary characters without having to think about
whether they are BMP or supplementary -- they're just characters™.

Among the new utility functions is `char'` (on analogy with clojure.core's `+'`
and other "enhanced" arithmetic operators that support arbitrary precision), an
extension of `clojure.core/char` that will return a string containing a
supplementary character if provided with a codepoint above U+FFFF,
e.g. `(char' 135641) => "𡇙"`"

Another convenient function is `char-range`, which returns the range (inclusive)
between two characters, e.g. `(char-range \a \z)` => `(\a \b \c ... \x \y \z)`.
This provides a concise, readable syntax for representing ranges of characters,
as compared to, e.g., `(map char (range (int \a) (inc (int \z))))`. As a bonus,
this function also supports supplementary characters, as it uses `char'`
internally.

My hope is that this library will end up in clojure.contrib or (my pipe dream)
as a part of Clojure proper as "clojure.char."

Any feedback and suggestions would be very welcome -- feel free to join the
[discussion](https://groups.google.com/forum/#!topic/clojure-dev/CVT5nqCz9XI)
going on the Clojure dev Google group.

Enjoy!

\- Dave Yarwood, 10/8/14

# To do:

* Remedy potential performance issues caused by dynamic type introspection,
[as noted by Mikera](https://groups.google.com/d/msg/clojure-dev/CVT5nqCz9XI/8oKdlmbOYk4J).


CHANGELOG.md
# CHANGELOG

## 0.2.1 (2019-05-21)

* `lower-case`, `upper-case`, and `title-case` now work on characters above the
  BMP range.

  Thanks again, @nibe!

## 0.2.0 (2019-05-20)

* Added a couple of convenience functions, `category-int` and `category`, which
  return the category of a character.

  Thanks, @nibe, for the contribution!

## 0.1.4 (2015-03-21)

* @solicode noticed an opportunity for a minor performance boost by avoiding using sets/`contains?`. Now using his `in?` macro, which expands the collection of things we're checking to a series of checking each individual thing for equality to the thing we're looking for.

* Added type hints, with minimal performance gains.

## 0.1.3 (2015-02-08)

* Enhanced the usability of char-range by making it behave more like clojure.core/range.
* Added comprehensive tests and benchmarks.


test/djy/char_test.clj
(ns djy.char-test
  (:require [clojure.test.check.generators :as gen]
            [clojure.test.check.properties :as prop]
            [clojure.test.check.clojure-test :refer (defspec)]
            [clojure.test :refer [deftest is]]
            [djy.char :as char]))

(def gen-bmp-char
  (gen/fmap char (gen/choose 0 65535)))

(def gen-stringified-bmp-char
  (gen/fmap str gen-bmp-char))

(def gen-high-surrogate
  (gen/fmap char (gen/choose 55296 56319)))

(def gen-low-surrogate
  (gen/fmap char (gen/choose 56320 57343)))

(def gen-supplementary-char
  (gen/fmap #(apply str %) (gen/tuple gen-high-surrogate gen-low-surrogate)))

(def gen-char
  (gen/fmap char/char' (gen/choose 0 1114111)))

(def gen-hebrew-char
  (gen/fmap char (gen/choose 1424 1535)))

(def gen-cherokee-char
  (gen/fmap char (gen/choose 5024 5119)))

(def gen-upper-case-letter
  (gen/fmap char (gen/choose (int \A) (int \Z))))

(def gen-lower-case-letter
  (gen/fmap char (gen/choose (int \a) (int \z))))

;;; testing code-point-of

(defspec code-point-of-an-integer-is-the-integer-itself
  (prop/for-all [code-point (gen/choose 0 1114111)]
    (= code-point
       (char/code-point-of code-point))))

(defspec code-point-of-a-bmp-character-is-its-code-point
  (prop/for-all [bmp-char gen-bmp-char]
    (= (int bmp-char)
       (char/code-point-of bmp-char))))

(defspec code-point-of-a-stringified-bmp-char-is-the-code-point-of-the-char
  (prop/for-all [str-bmp-char gen-stringified-bmp-char]
    (= (int (first str-bmp-char))
       (char/code-point-of str-bmp-char))))

(defspec code-point-of-a-supplementary-char-is-the-code-point-at-index-0
  (prop/for-all [supplementary-char gen-supplementary-char]
    (= (.codePointAt supplementary-char 0)
       (char/code-point-of supplementary-char))))

;;; testing char'

(defspec code-points-0-65535-are-chars
  (prop/for-all [bmp-code-point (gen/choose 0 65535)]
    (char? (char/char' bmp-code-point))))

(defspec code-points-65536-1114111-are-strings
  (prop/for-all [supplementary-code-point (gen/choose 65536 1114111)]
    (string? (char/char' supplementary-code-point))))

(defspec code-point-of-and-char'-are-inverse-operations
  (prop/for-all [code-point (gen/choose 0 1114111)]
    (= code-point (char/code-point-of (char/char' code-point)))))

;;; testing char comparator functions

(defspec char-less-than-or-equal-to
  (prop/for-all [characters (gen/such-that not-empty (gen/vector gen-char))]
    (apply char/char<= (sort-by char/code-point-of characters))))

(defspec char-less-than
  (prop/for-all [characters (gen/such-that #(and (not-empty %)
                                            (= (count %) (count (distinct %))))
                              (gen/vector gen-char))]
    (apply char/char< (sort-by char/code-point-of characters))))

;;; testing char/prev

(defspec char-prev
  (prop/for-all [characters (gen/bind
                              (gen/such-that #(> (char/code-point-of %) 0)
                                gen-char)
                              #(gen/tuple
                                 (gen/return %)
                                 (gen/return (char/prev %))))]
    (= (char/code-point-of (first characters))
       (inc (char/code-point-of (second characters))))))

;;; testing char-range

(def gen-some-chars
  "A subset of characters spanning the BMP and supplementary ranges.
   Using this instead of gen-char so the test won't spend an eternity
   trying to build a huge character range if the characters happen to
   be far apart (which is very probable)."
  (gen/fmap char/char' (gen/choose 50000 70000)))

(defspec char-range-is-like-range-for-characters
  (prop/for-all [[a b] (gen/such-that #(>= (count %) 2)
                         (gen/vector gen-some-chars))]
    (= (range (char/code-point-of a) (inc (char/code-point-of b)))
       (map char/code-point-of (char/char-range a b)))))

;;; testing char-seq

(defspec char-seq-provides-accurate-character-counts
  (prop/for-all [characters (gen/vector gen-char)]
    (= (count characters)
       (count (char/char-seq (apply str characters))))))

;;; testing surrogates

(defspec surrogates-constitute-a-character
  (prop/for-all [character gen-char]
    (= (str character)
       (apply str (char/surrogates character)))))

;;; testing unicode-block-of

(defspec unicode-block-of-alphanumeric-chars-is-basic-latin
  (prop/for-all [alphanumeric-char gen/char-alphanumeric]
    (= (char/unicode-block-of alphanumeric-char) "BASIC_LATIN")))

(defspec unicode-block-of-hebrew-chars-is-hebrew
  (prop/for-all [hebrew-char gen-hebrew-char]
    (= (char/unicode-block-of hebrew-char) "HEBREW")))

(defspec unicode-block-of-cherokee-chars-is-cherokee
  (prop/for-all [cherokee-char gen-cherokee-char]
    (= (char/unicode-block-of cherokee-char) "CHEROKEE")))

(defspec all-chars-belong-to-a-unicode-block
  (prop/for-all [character gen-char]
    (string? (char/unicode-block-of character))))

;;; testing category

(defspec category-lower-case-letter
  (prop/for-all [lower-case-letter gen-lower-case-letter]
    (= (char/category lower-case-letter)
       "Ll")))

(defspec category-upper-case-letter
  (prop/for-all [upper-case-letter gen-upper-case-letter]
    (= (char/category upper-case-letter)
       "Lu")))

;;; testing boolean functions

(defspec defined?
  (prop/for-all [character gen-char]
    (contains? #{true false} (char/defined? character))))

(defspec bmp?
  (prop/for-all [bmp-char gen-bmp-char]
    (char/bmp? bmp-char)))

(defspec supplementary?
  (prop/for-all [supplementary-char gen-supplementary-char]
    (char/supplementary? supplementary-char)))

(defspec high-surrogate?
  (prop/for-all [high-surrogate gen-high-surrogate]
    (char/high-surrogate? high-surrogate)))

(defspec low-surrogate?
  (prop/for-all [low-surrogate gen-low-surrogate]
    (char/low-surrogate? low-surrogate)))

(defspec surrogate?
  (prop/for-all [surrogate (gen/one-of [gen-high-surrogate gen-low-surrogate])]
    (char/surrogate? surrogate)))

(defspec letter-or-digit?
  (prop/for-all [character gen/char-alphanumeric]
    (char/letter-or-digit? character)))

;;; testing case conversion functions

(defspec lower-case
  (prop/for-all [upper-case-letter gen-upper-case-letter]
    (= (char/lower-case upper-case-letter)
       (Character/toLowerCase upper-case-letter))))

(defspec upper-case
  (prop/for-all [lower-case-letter gen-lower-case-letter]
    (= (char/upper-case lower-case-letter)
       (Character/toUpperCase lower-case-letter))))

(deftest lower-case-supplementary
  (is (= (char/lower-case "𐐄")
         "𐐬"))
  (is (= (char/lower-case "😀")
         "😀")))

(deftest upper-case-supplementary
  (is (= (char/upper-case "𐐬")
         "𐐄"))
  (is (= (char/upper-case "😀")
         "😀")))


benchmark/djy/char_benchmark.clj
(ns djy.char-benchmark
  (:require [criterium.core :refer :all]
            [djy.char :as char :refer (char')]))

(defn bmp-code-points
  "Generates n random code points in the BMP range."
  [n]
  (repeatedly n #(rand-int 65536)))

(defn supp-code-points
  "Generates n random code points in the supplementary range."
  [n]
  (let [supp-range (range 65536 1114112)]
    (repeatedly n #(rand-nth supp-range))))

(defn code-points
  "Generates n random code points."
  [n]
  (repeatedly n #(rand-int 1114112)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking conversion of ints to chars with char'...\n")

(let [cp (code-points 1000)]
  (quick-bench (doall (map char' cp)) :verbose))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking conversion of BMP ints to chars with char'...\n")

(let [cp (bmp-code-points 1000)]
  (quick-bench (doall (map char' cp)) :verbose))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking conversion of BMP ints to chars with clojure.core/char...\n")

(let [cp (bmp-code-points 1000)]
  (quick-bench (doall (map char cp)) :verbose))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking conversion of supplementary ints to chars with char'...\n")

(let [cp (supp-code-points 1000)]
  (quick-bench (doall (map char' cp)) :verbose))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking conversion of supplementary ints to chars with Java inter-op...\n")

(let [cp (supp-code-points 1000)]
  (quick-bench (doall (map #(apply str (Character/toChars %)) cp)) :verbose))


benchmark/djy/char_range_benchmark.clj
(ns djy.char-range-benchmark
  (:require [criterium.core :refer :all]
            [djy.char :as char :refer (char-range)]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking char-range for BMP characters...\n")

(quick-bench (doall (char-range 10000 60000)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking mapping clojure.core/char over the same range...\n")

(quick-bench (doall (map char (range 10000 60001))) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking char-range for supplementary characters...\n")

(quick-bench (doall (char-range 100000 150000)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking the equivalent Java interop for the same range...\n")

(quick-bench (doall (map #(apply str (Character/toChars %))
                         (range 100000 150000)))
             :verbose)


benchmark/djy/text_processing_benchmark.clj
(ns djy.text-processing-benchmark
  (:require [criterium.core :refer :all]
            [djy.char :as char :refer (char-seq code-point-of)]
            [clojure.java.io :as io]))

(def i18nguy
  "Source of a website containing 124 supplementary characters.
   Roughly 15,000 total characters in length."
  (slurp "http://www.i18nguy.com/unicode/supplementary-test.html"))

(def bmp-chars-string
  (slurp (io/resource "sample/bmp-chars.txt")))

(def supplementary-chars-string
  (slurp (io/resource "sample/supplementary-chars.txt")))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println (str "\nBenchmarking char-seq on the source of\n"
              "http://www.i18nguy.com/unicode/supplementary-test.html...\n"))

(quick-bench (doall (char-seq i18nguy)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking clojure.core/seq on the same text...\n")

(quick-bench (doall (seq i18nguy)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking code-point-of on the same text...\n")

(quick-bench (doall (map code-point-of i18nguy)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking char-seq on a string of all BMP characters...\n")

(quick-bench (doall (char-seq bmp-chars-string)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking clojure.core/seq on the same text...\n")

(quick-bench (doall (seq bmp-chars-string)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking code-point-of on the same text...\n")

(quick-bench (doall (map code-point-of bmp-chars-string)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println (str "\nBenchmarking char-seq on a string of the first 65,536 "
              "supplementary characters...\n"))

(quick-bench (doall (char-seq supplementary-chars-string)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking clojure.core/seq on the same text...\n")

(quick-bench (doall (seq supplementary-chars-string)) :verbose)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(println "\nBenchmarking code-point-of on the same text...\n")

(quick-bench (doall (map code-point-of supplementary-chars-string)) :verbose)


src/djy/char.clj
(ns ^{:doc "A library of character-related utility functions for Clojure.

        Note: There are several functions in this library that may collide with other
        namespaces. Recommended usage is:
            (ns your.namespace
              (:require [djy.char :as char :refer (...)]))

        Many of these functions are polymorphic in that they can accept as an argument
        either a character, an integer representing a Unicode code point, or a string.
        If the argument is a string, the function uses the Unicode code point found at
        index 0 of the string, and the rest of the string (if the string happens to be
        longer than one Unicode code point) is ignored, e.g.:

            (code-point-of '\ud800\udc00') ;=> 65536
            (code-point-of 'hello')        ;=> 104 (the code point of 'h')
            [imagine that the single quotes above are double quotes]

        When dealing with supplementary Unicode characters with code points higher than
        U+FFFF, the character must be represented either as a String (e.g. '\ud800\udc00')
        or as a code point (e.g. 65536). This is due to the inherent limitations of 16-
        bit Java characters and the fact that Clojure character literals are built on top
        of them.

        Glossary:

        * Code Point: any value in the Unicode codespace; that is, the range of integers
             from 0 to 10FFFF (in decimal: 1114111). Not all code points are assigned to
             encoded characters.
        * Basic Multilingual Plane (BMP): Unicode characters in the range 0000-FFFF.
        * Supplementary characters: Unicode characters with code points > FFFF. Java
              represents these characters as a string containing two 16-bit characters,
              a high-surrogate and a low-surrogate (in that order).
        * High-surrogates: characters in the range D800-DBFF
        * Low-surrogates: characters in the range DC00-DFFF
        * ASCII: characters in the range 0000-007F (0-127)
        * Latin-1: characters in the range 0080-00FF (0-255)"
      :author "Dave Yarwood"}
  djy.char
  (:require [clojure.string :as str])
  (:refer-clojure :exclude [next symbol?]))

(defmacro in? [coll val]
  (let [val-sym (gensym)]
    `(let [~val-sym ~val]
       (or ~@(for [x coll] `(= ~val-sym ~x))))))

;;; Utility functions ;;;

(defprotocol HasCodePoint
  (code-point-of [this] "Returns the Unicode code point of a character."))

(extend-protocol HasCodePoint
  java.lang.Number
    (code-point-of [n] {:pre [(<= 0 n 1114111)]} n)
  java.lang.Character
    (code-point-of [ch] (int ch))
  java.lang.String
    (code-point-of [s] {:pre [(not (empty? s))]} (.codePointAt s 0)))

(defn char'
  "Like clojure.core/char, returns the character at a given code point.

   Whereas char will throw an error if the code point is greater than 65535 (U+FFFF),
   char' will return the supplemental character (in string form) at that code point."
  {:added "1.6"}
  [n]
  {:pre [(<= 0 n 1114111)]}
  (try
    (char n)
    (catch IllegalArgumentException e
      (apply str (Character/toChars n)))))

(defn char<
  "Like clojure.core/<, but it converts its arguments (which are each expected to be
   a character, a code point, or a string containing a supplementary character) to code
   points. Useful for determining whether a collection of characters is in order by code
   point."
  {:added "1.6"}
  [& chs]
  (apply < (map code-point-of chs)))

(defn char>
  "Like clojure.core/>, but it converts its arguments (which are each expected to be
   a character, a code point, or a string containing a supplementary character) to code
   points. Useful for determining whether a collection of characters is in order by code
   point."
  {:added "1.6"}
  [& chs]
  (apply > (map code-point-of chs)))

(defn char<=
  "Like clojure.core/<=, but it converts its arguments (which are each expected to be
   a character, a code point, or a string containing a supplementary character) to code
   points. Useful for determining whether a collection of characters is in order by code
   point."
  {:added "1.6"}
  [& chs]
  (apply <= (map code-point-of chs)))

(defn char>=
  "Like clojure.core/<=, but it converts its arguments (which are each expected to be
   a character, a code point, or a string containing a supplementary character) to code
   points. Useful for determining whether a collection of characters is in order by code
   point."
  {:added "1.6"}
  [& chs]
  (apply >= (map code-point-of chs)))

(defn prev
  "Returns the character before the one given, i.e. the character whose code point is one
   less than the one given. Returns a string if the resulting character is supplementary."
  {:added "1.6"}
  [ch]
  (char' (dec (code-point-of ch))))

(defn next
  "Returns the character after the one given, i.e. the character whose code point is one
   greater than the one given. Returns a string if the resulting character is supplementary."
  {:added "1.6"}
  [ch]
  (char' (inc (code-point-of ch))))

(defn char-range
  "Given two characters or code points, returns the range (inclusive) between them.
   e.g. (char-range a z) => (a b c d e ... x y z) [imagine these are all characters]
   Represents supplementary characters as strings.

   As this function uses clojure.core/range internally, the result is a lazy seq.

   Emulates the behavior of clojure.core/range when given different numbers of
   arguments. For example, an optional third argument can be provided as the step
   (defaults to 1)."
  {:added "1.6"}
  ([]
    (cons (char 0) (lazy-seq (map (comp char' inc) (range)))))
  ([end]
    (char-range 0 end 1))
  ([start end]
    (char-range start end 1))
  ([start end step]
    (map char' (range (code-point-of start) (inc (code-point-of end)) step))))

(declare supplementary?)

(defn char-seq
  "Generates a lazy seq of characters from a string, appropriately handling
   supplementary characters by representing them as strings. Equivalent to (seq s),
   except that supplementary characters are represented in string form rather than
   broken up into their surrogate characters."
  {:added "1.6"}
  [^String s]
  {:pre [(string? s)]}
  (when (seq s)
    (let [supplementary? (supplementary? (code-point-of s))]
      (cons (if supplementary? (subs s 0 2) (first s))
            (lazy-seq (char-seq (subs s (if supplementary? 2 1))))))))

(defn surrogates
  "Returns a seq containing the surrogate pair (high, low) for a supplementary character
   (in string form) or code point. If given a BMP character or code point, returns a seq
   containing just that character."
  {:added "1.6"}
  [ch]
  (seq (Character/toChars (code-point-of ch))))

(defn unicode-block-of
  "Returns the name of the Unicode block for the character or code point."
  {:added "1.6"}
  [ch]
  (str (java.lang.Character$UnicodeBlock/of ^long (code-point-of ch))))

(comment
  "This function requires JDK >= 1.7"
  (defn name
    "Returns the Unicode name of the character or code point, or nil if the code point is
     unassigned."
    {:added "1.6"}
    [ch]
    (Character/getName (code-point-of ch)))
)

(defn category-int
  "Returns an integer value indicating the general category of the given character or code
   point."
  {:added "1.6"}
  [ch]
  (Character/getType ^long (code-point-of ch)))

(def category-names
  "A map where general category names are indexed by the associated integer constants."
  {Character/UPPERCASE_LETTER "Lu"
   Character/LOWERCASE_LETTER "Ll"
   Character/TITLECASE_LETTER "Lt"
   Character/MODIFIER_LETTER "Lm"
   Character/OTHER_LETTER "Lo"

   Character/NON_SPACING_MARK "Mn"
   Character/COMBINING_SPACING_MARK "Mc"
   Character/ENCLOSING_MARK "Me"

   Character/DECIMAL_DIGIT_NUMBER "Nd"
   Character/LETTER_NUMBER "Nl"
   Character/OTHER_NUMBER "No"

   Character/CONNECTOR_PUNCTUATION "Pc"
   Character/DASH_PUNCTUATION "Pd"
   Character/START_PUNCTUATION "Ps"
   Character/END_PUNCTUATION "Pe"
   Character/INITIAL_QUOTE_PUNCTUATION "Pi"
   Character/FINAL_QUOTE_PUNCTUATION "Pf"
   Character/OTHER_PUNCTUATION "Po"

   Character/MATH_SYMBOL "Sm"
   Character/CURRENCY_SYMBOL "Sc"
   Character/MODIFIER_SYMBOL "Sk"
   Character/OTHER_SYMBOL "So"

   Character/SPACE_SEPARATOR "Zs"
   Character/LINE_SEPARATOR "Zl"
   Character/PARAGRAPH_SEPARATOR "Zp"

   Character/CONTROL "Cc"
   Character/FORMAT "Cf"
   Character/SURROGATE "Cs"
   Character/PRIVATE_USE "Co"
   Character/UNASSIGNED "Cn"})

(defn category
  "Returns the name of the general category of the given character or code point."
  {:added "1.6"}
  [ch]
  (get category-names (category-int ch)))

;;; Boolean functions ;;;

(defn defined?
  "Determines whether a character or code point is defined in Unicode, i.e. it has an entry
   in the UnicodeData file, or has a value in a range defined by the UnicodeData file."
  {:added "1.6"}
  [ch]
  (Character/isDefined ^long (code-point-of ch)))

(defn bmp?
  "Determines whether a character or code point is in the Basic Multilingual Plane (BMP),
   which includes characters U+0000 through U+FFFF."
  {:added "1.6"}
  [ch]
  (<= 0 (code-point-of ch) 65535))

(defn supplementary?
  "Determines whether a character or code point is a supplementary character,
   i.e. code point in the range U+10000 through U+10FFFF."
  {:added "1.6"}
  [ch]
  (<= 65536 (code-point-of ch) 1114111))

(defn high-surrogate?
  "Determines whether a character or code point is in the high-surrogates range,
   i.e. U+D800 through U+DBFF."
  {:added "1.6"}
  [ch]
  (<= 55296 (code-point-of ch) 56319))

(defn low-surrogate?
  "Determines whether a character or code point is in the low-surrogates range,
   i.e. U+DC00 through U+DFFF."
  {:added "1.6"}
  [ch]
  (<= 56320 (code-point-of ch) 57343))

(defn surrogate?
  "Determines whether or not a character/code point is a high- or low-surrogate."
  {:added "1.6"}
  [ch]
  (let [cp (code-point-of ch)]
    (or
      (<= 55296 cp 56319)
      (<= 56320 cp 57343))))

(defn ascii?
  "Determines whether a character or code point is an ASCII character."
  {:added "1.6"}
  [ch]
  (<= 0 (code-point-of ch) 127))

(defn latin-1?
  "Determines whether a character or code point is in the Latin-1 charset."
  {:added "1.6"}
  [ch]
  (<= 0 (code-point-of ch) 255))

(defn control?
  "Determines whether a character or code point is an ISO control character."
  {:added "1.6"}
  [ch]
  (let [cp (code-point-of ch)]
    (or
      (<= 0 cp 31)
      (<= 127 cp 159))))

(comment
  "This function requires JDK >= 1.7"
  (defn alphabetic?
    "Determines whether a character or code point is Alphabetic, as defined by the
     Unicode standard."
    {:added "1.6"}
    [ch]
    (Character/isAlphabetic (code-point-of ch)))
)

(defn letter?
  "Determines whether a character or code point is a letter."
  {:added "1.6"}
  [ch]
  (Character/isLetter ^long (code-point-of ch)))

(defn digit?
  "Determines whether a character or code point is a digit. A character is a digit if its
   general category type is DECIMAL_DIGIT_NUMBER; this includes many ranges of characters
   including ISO-LATIN-1 digits, full-width digits, Devanagari digits, etc."
  {:added "1.6"}
  [ch]
  (Character/isDigit ^long (code-point-of ch)))

(defn letter-or-digit?
  "Determines whether a character or code point is a letter or digit."
  {:added "1.6"}
  [ch]
  (let [cp (code-point-of ch)]
    (or (Character/isLetter ^long cp) (Character/isDigit ^long cp))))

(defn oct-digit?
  "Determines whether a character or code point is an octal digit (0-7).
   Returns true only for ISO-LATIN-1 digits."
  {:added "1.6"}
  [ch]
  (<= 48 (code-point-of ch) 55))

(defn hex-digit?
  "Determines whether a character or code point is a hexadecimal digit.
   Returns true only for ISO-LATIN-1 characters."
  {:added "1.6"}
  [ch]
  (let [cp (code-point-of ch)]
    (or
      (<= 48 cp 57)     ; 0-9
      (<= 65 cp 70)     ; A-F
      (<= 97 cp 102)))) ; a-f

(comment
  "This function requires JDK >= 1.7"
  (defn ideograph?
    "Determines whether a character or code point is a CKJV ideograph."
    {:added "1.6"}
    [ch]
    (Character/isIdeographic (code-point-of ch)))
)

(defn whitespace?
  "Determines whether a character or code point is a Java whitespace character."
  {:added "1.6"}
  [ch]
  (Character/isWhitespace ^long (code-point-of ch)))

(defn punctuation?
  "Determines whether a character or code point is a punctuation character, according to
   the Unicode standard."
  {:added "1.6"}
  [ch]
  (in? #{Character/CONNECTOR_PUNCTUATION, Character/DASH_PUNCTUATION,
         Character/START_PUNCTUATION, Character/END_PUNCTUATION,
         Character/INITIAL_QUOTE_PUNCTUATION, Character/FINAL_QUOTE_PUNCTUATION,
         Character/OTHER_PUNCTUATION}
       (category-int ch)))

(defn mark?
  "Determines whether a character or code point is a mark character, according to the
   Unicode standard."
  {:added "1.6"}
  [ch]
  (in? #{Character/COMBINING_SPACING_MARK, Character/ENCLOSING_MARK,
         Character/NON_SPACING_MARK}
       (category-int ch)))

(defn symbol?
  "Determines whether a character or code point is a symbol character, according to the
   Unicode standard."
  {:added "1.6"}
  [ch]
  (in? #{Character/MATH_SYMBOL, Character/CURRENCY_SYMBOL,
         Character/MODIFIER_SYMBOL, Character/OTHER_SYMBOL}
       (category-int ch)))

(defn separator?
  "Determines whether a character or code point is a separator, according to the Unicode
   standard."
  {:added "1.6"}
  [ch]
  (in? #{Character/LINE_SEPARATOR, Character/PARAGRAPH_SEPARATOR,
         Character/SPACE_SEPARATOR}
       (category-int ch)))


(defn lower-case?
  "Determines whether a character or code point is a lower-case letter."
  {:added "1.6"}
  [ch]
  (Character/isLowerCase ^long (code-point-of ch)))

(defn upper-case?
  "Determines whether a character or code point is an upper-case letter."
  {:added "1.6"}
  [ch]
  (Character/isUpperCase ^long (code-point-of ch)))

(defn title-case?
  "Determines whether a character or code point is a title-case letter."
  {:added "1.6"}
  [ch]
  (Character/isTitleCase ^long (code-point-of ch)))

;;; Case conversion functions ;;;

(defn lower-case
  "Returns the lower-case counterpart of the given character or code point."
  {:added "1.6"}
  [ch]
  (char' (Character/toLowerCase ^long (code-point-of ch))))

(defn upper-case
  "Returns the upper-case counterpart of the given character or code point."
  {:added "1.6"}
  [ch]
  (char' (Character/toUpperCase ^long (code-point-of ch))))

(defn title-case
  "Returns the title-case counterpart of the given character or code point."
  {:added "1.6"}
  [ch]
  (char' (Character/toTitleCase ^long (code-point-of ch))))
